llzk_backend/src/codegen.rs

```diff
@@ -1,62 +1,156 @@
+use ansi_term::Color;
+use anyhow::{anyhow, Result};
+use llzk::{
+    error::Error,
+    prelude::{
+        felt, function,
+        r#struct::{
+            self,
+            helpers::{compute_fn, constrain_fn},
+        },
+        undef, ArrayType, FeltConstAttribute, FeltType, FuncDefOp, FuncDefOpLike, FuncDefOpRef,
+        FuncDefOpRefMut, FunctionType, IntegerAttribute, LlzkContext, OperationMutLike,
+        PublicAttribute, StructDefOp, StructDefOpLike, StructDefOpRef, StructDefOpRefMut,
+        StructType,
+    },
+};
+use melior::ir::{
+    operation::{OperationLike as _, OperationRefMut, WalkOrder, WalkResult},
+    Attribute, Block, BlockLike, BlockRef, Identifier, Location, Module, Operation, OperationRef,
+    RegionLike, Type, Value,
+};
+use num_bigint_dig::BigInt;
+use num_traits::cast::ToPrimitive;
+use program_structure::{
+    ast::{AssignOp, Expression, Meta, SignalType, Statement, VariableType},
+    error_code::ReportCode,
+    error_definition::Report,
+    file_definition::{FileID, FileLocation},
+    function_data::FunctionData,
+    program_archive::ProgramArchive,
+    template_data::TemplateData,
+};
 use std::{
+    collections::HashMap,
+    convert::{TryFrom, TryInto as _},
     fs::{self, File},
     io::Write,
+    ops::{Deref, DerefMut},
     os::raw::c_void,
     path::Path,
+    slice,
 };
-use ansi_term::Color;
-use anyhow::Result;
-use program_structure::{
-    file_definition::{FileID, FileLibrary, FileLocation},
-    program_archive::ProgramArchive,
-};
-use melior::ir::{operation::OperationLike as _, Location, Module, ValueLike};
-use llzk::prelude::LlzkContext;
 
-/// Stores necessary context for generating LLZK IR along with the generated `Module`.
+/// Stack of blocks where the top block is the current block where code should be appended and the
+/// previous block in the list is the parent of the block after it. When an op containing nested
+/// blocks is encountered, the current block within that op is pushed to the stack so that any code
+/// generated will be placed inside that block and when the nested block is complete, it is popped.
+struct BlockContextStack<'llzk> {
+    /// The function entry block.
+    initial_block: BlockRef<'llzk, 'llzk>,
+    /// Additional nesting of blocks within the function representing the current insertion point.
+    other_blocks: Vec<BlockRef<'llzk, 'llzk>>,
+}
+
+impl<'llzk> BlockContextStack<'llzk> {
+    fn push(&mut self, item: BlockRef<'llzk, 'llzk>) {
+        self.other_blocks.push(item);
```

Comment:
Using the lifetimes that way means that the block must live for the whole lifetime of the context, which may freak out the borrow checker at times. I'd do this instead:

```suggestion
struct BlockContextStack<'llzk,'blk: 'llzk> {
    /// The function entry block.
    initial_block: BlockRef<'llzk, 'blk>,
    /// Additional nesting of blocks within the function representing the current insertion point.
    other_blocks: Vec<BlockRef<'llzk, 'blk>>,
}

impl<'llzk, 'blk: 'llzk> BlockContextStack<'llzk, 'blk> {
    fn push(&mut self, item: BlockRef<'llzk, 'blk>) {
        self.other_blocks.push(item);
```